### PR TITLE
DeepSeek V3 FSDP Fix for Precision-Aware Optimizer

### PR DESCRIPTION
**Pull request by tomlifu**
# What does this PR do ?
This PR adds fix for precision-aware optimizer for DeepSeek V3.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Review by shjwudp — APPROVED** [review]
LGTM, thanks!

**Review by BoxiangW — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 1]
```diff
@@ -2502,9 +2499,6 @@ def update_main_grads(self):
             # that a precision-aware optimizer is used.
             if getattr(self, "use_precision_aware_optimizer", False):
                 setattr(param, "decoupled_grad", grad)
-            else:
-                # Attach the gradient to the optimizer parameter.
-                setattr(param, "grad", grad.to(param.dtype) if grad is not None else None)
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2507]: Not all optimizers are precision-aware, if you don't attach the gradient to the parameter normal optimizers will not work.

This should fail unit tests, have they run yet? Why don't we just have an argument that satisfies `getattr(self, "use_precision_aware_optimizer", False)` instead if you want to use this? You're welcome to add arguments to `MegatronFSDP.__init__()`!

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 2]
```diff
@@ -2473,9 +2473,6 @@ def update_main_grads(self):
             optimizer_grad = group.main_grad_buffer.get_item(
                 item_id, only_shard=sharded_optimizer_state
             )
-            if group.main_weight_buffer is not None:
-                # Convert the gradient to the main weight buffer dtype.
-                optimizer_grad = optimizer_grad.to(param.dtype)
```
Comment (cspades) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2478]: This is important for when your gradient dtype doesn't match the main weight dtype. (`preserve_fp32_weights=True` and `reduce_grads_in_fp32=False` and vice versa, which is our current low-effort tentative solution for mixed-precision buffers.) Why are we deleting this?

**Review by cspades — CHANGES_REQUESTED** [review]
Some really concerning changes in this PR that breaks OSS compatibility. Can we add arguments that allow the user to set these options instead of deleting code?

@tomlifu This looks like a hotfix for a specific model, am I right? You can add switches to control this behavior, I was planning to add something to expose the precision-aware optimizer hacks to the API as well!

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 1 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2507]: Yup, the reason the `test_mfsdp_fully_shard` unit test hangs is directly because of this:
```
            if getattr(self, "use_precision_aware_optimizer", False):
                setattr(param, "decoupled_grad", grad)
            # else:
            #     # Attach the gradient to the optimizer parameter.
            #     setattr(param, "grad", grad.to(param.dtype) if grad is not None else None)
```
I just reproduced it. We can't delete this, I can help add an argument like `MegatronFSDP.__init__(use_precision_aware_optimizer: bool = False)` which will control this behavior. WDYT @tomlifu @shjwudp ?

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2507]: The hanging CI unit test is also my bad. (Funny thing is when you have 1 year of PyTorch experience, code you wrote 4 months ago is pretty crappy. 🙏🏻) I thought I put it on my backlog to fix this, but for some reason I only gather to Rank 0, so if the test fails:
```
> /opt/megatron-lm/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py(346)test_fully_shard()
-> if sharding_group.rank() == 0:
(Pdb) FAILED
```
It only fails on Rank 0 while the other 7 ranks continue and likely wait for the terminated Rank 0 on a subsequent unit test, leading to a timeout, which isn't an acceptable turnaround time for a unit test. (Sorry again!)

I should just simply use https://docs.pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather_object and fail on all ranks.

Update: This works. Fails immediately when it should, succeeds when I put the `setattr(param, "grad", ...)` code back.
```
diff --git a/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py b/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py
index ee485dd0b0..195caf81ea 100644
--- a/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py
+++ b/tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py
@@ -326,27 +326,21 @@ class TestMegatronFsdpFullyShard:
                 # Because of uneven sharding, we need to gather the result from all ranks
                 # to verify if any gradients exist or not at this step of training.
                 grads_exist_gathered = [None] * sharding_group.size()
-                torch.distributed.gather_object(
-                    grads_exist,
-                    object_gather_list=grads_exist_gathered if sharding_group.rank() == 0 else None,
+                torch.distributed.all_gather_object(
+                    object_list=grads_exist_gathered,
+                    obj=grads_exist,
                     group=sharding_group,
-                    group_dst=0,
                 )
-                if sharding_group.rank() == 0:
-                    # Gradients exist on at least one of the optimizer sharding ranks.
-                    # Update grads_exist on Rank 0 only.
-                    grads_exist = any(grads_exist_gathered)
-                torch.distributed.barrier()
+                # Gradients exist on at least one of the optimizer sharding ranks.
+                grads_exist = any(grads_exist_gathered)
 
             # Gradients do not exist until synchronization is activated.
-            # Use collected result on Rank 0 only.
-            if sharding_group.rank() == 0:
-                if step == NUM_STEPS - 1:
-                    assert grads_exist, "Root module gradients should exist on final microbatch."
-                else:
-                    assert (
-                        not grads_exist
-                    ), "Root module gradients should not exist prior to optimization step."
+            if step == NUM_STEPS - 1:
+                assert grads_exist, "Root module gradients should exist on final microbatch."
+            else:
+                assert (
+                    not grads_exist
+                ), "Root module gradients should not exist prior to optimization step."
             torch.distributed.barrier()
```

Either way, test hang did its job, don't delete this code!

Comment (shjwudp) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2507]: Yes, @tomlifu , your changes should only affect the code where `use_precision_aware_optimizer=True`.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 2 above)
Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2478]: I added this back with a case when use_precision_aware_optimizer=false because a fp32 param and bf16 gradient pair can be provided to the precision-aware optimizer. In this case, we shouldn’t forcibly cast the gradient to the same data type.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 3]
```diff
@@ -2500,7 +2501,7 @@ def update_main_grads(self):
 
             # The presence of main_grad_buffer but no main_weight_buffer may imply
             # that a precision-aware optimizer is used.
-            if getattr(self, "use_precision_aware_optimizer", False):
+            if getattr(self, "use_precision_aware_optimizer", True):
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Why do we change default setting here?

Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: we shouldn’t forcibly cast the gradient to the same data type as the parameter when use_precision_aware_optimizer=True to match with my changes here: https://github.com/NVIDIA/Megatron-LM/pull/2204/files#r2540423699

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 2 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2478]: Thanks for adding it back!

If you want to avoid doing this cast operation when using a precision-aware optimizer, shouldn't it be `not getattr(self, "use_precision_aware_optimizer", False)`:
```
if not getattr(self, "use_precision_aware_optimizer", False):
    # Convert the gradient to the main weight buffer dtype if *not* using a precision-aware optimizer.
    optimizer_grad = optimizer_grad.to(param.dtype)
```

So if we are **_not_** using a precision-aware optimizer, then we **_should_** make sure our gradient and parameter buffer dtypes are the same, otherwise we'll hit a dtype incompatibility error.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 3 above)
Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Shouldn't this also be `False` by default? Most optimizers are not precision-aware, and I think you should manually set this attribute when you are using a precision-aware optimizer.

So it should be:
- If using precision-aware optimizer, use `decoupled_grad` and do not cast.
- If not using precision-aware optimizer, use `grad` and cast.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py [hunk 4]
```diff
@@ -772,6 +772,8 @@ def register_submesh(device_mesh, submesh, is_expert_parallel):
 
         # Register EP submeshes
         if self.expt_device_mesh is not None:
+            register_submesh(self.expt_device_mesh, hsdp_submesh, True)
```
Comment (kunlunl) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py]: Is this related to precision-aware optimizer? If not, maybe better to create another MR?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 5]
```diff
@@ -2474,8 +2474,9 @@ def update_main_grads(self):
                 item_id, only_shard=sharded_optimizer_state
             )
             if group.main_weight_buffer is not None:
-                # Convert the gradient to the main weight buffer dtype.
-                optimizer_grad = optimizer_grad.to(param.dtype)
+                if getattr(self, "use_precision_aware_optimizer", False):
```
Comment (kunlunl) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Could you explain this change?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 2 above)
Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py:2478]: yes, it should be `if not getattr(self, "use_precision_aware_optimizer", False):`.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 3 above)
Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: The default should be false, and I changed it back.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 5 above)
Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: We'd like to avoid doing this cast operation when using a precision-aware optimizer, it should be `not getattr(self, "use_precision_aware_optimizer", False):`

File: megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py (hunk 4 above)
Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py]: Sure, I'll create another PR for this.

Comment (tomlifu) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/utils.py]: https://github.com/NVIDIA/Megatron-LM/pull/2388

**Comment by cspades** [comment]
/ok to test [bd03c19](https://github.com/NVIDIA/Megatron-LM/pull/2204/commits/bd03c19d79cb57afefb96e84625c474631e7b7e1)

**Comment by shjwudp** [comment]
@tomlifu Every PR submitted to the dev branch needs a corresponding mirror MR on the main branch. Please also create a mirror MR on the main branch. Thank you!

**Comment by tomlifu** [comment]
> @tomlifu Every PR submitted to the dev branch needs a corresponding mirror MR on the main branch. Please also create a mirror MR on the main branch. Thank you!

I created a mirror PR to main here: https://github.com/NVIDIA/Megatron-LM/pull/2466
